client/websocket_client.tsx

```diff
@@ -132,7 +154,35 @@ export default class WebSocketClient {
                     Reflect.deleteProperty(this.responseCallbacks, msg.seq_reply);
                 }
             } else if (this.eventCallback) {
-                if (msg.seq !== this.serverSequence && this.missedEventCallback) {
+                if (reliableWebSockets) {
+                    // We check the hello packet, which is always the first packet in a stream.
+                    if (msg.event === SocketEvents.HELLO && this.missedEventCallback) {
+                        console.log('got connection id ', msg.data.connection_id); //eslint-disable-line no-console
+                        // If we already have a connectionId present, and server sends a different one,
+                        // that means it's either a long timeout, or server restart, or sequence number is not found.
+                        // Then we do the sync calls, and reset sequence number to 0.
+                        if (this.connectionId !== '' && this.connectionId !== msg.data.connection_id) {
+                            console.log('long timeout, or server restart, or sequence number is not found.'); //eslint-disable-line no-console
+                            this.missedEventCallback();
+                            this.serverSequence = 0;
+                        }
+
+                        // If it's a fresh connection, we have to set the connectionId regardless.
+                        // And if it's an existing connection, setting it again is harmless, and keeps the code simple.
+                        this.connectionId = msg.data.connection_id;
+                    }
+
+                    // Now we check for sequence number, and if it does not match,
+                    // we just disconnect and reconnect.
+                    if (msg.seq !== this.serverSequence) {
+                        console.log('missed websocket event, act_seq=' + msg.seq + ' exp_seq=' + this.serverSequence); //eslint-disable-line no-console
+                        // We are not calling this.close() because we need to auto-restart.
+                        this.connectFailCount = 0;
+                        this.responseSequence = 1;
+                        this.conn?.close(); // Will auto-reconnect after MIN_WEBSOCKET_RETRY_TIME.
```

Comment:
How will closing and reopening the websocket let the client get back in sync? Does this cause the server to resend any missed events? It also looks like we're not calling the old reconnect logic anywhere either which I thought we needed in the case where there was a longer disconnect.

Reply:
Oh, I see. I didn't realize we already had handling for missed events from back when the sequence numbers were originally added